eth/borfinality/whitelist/finality.go

```diff
@@ -0,0 +1,96 @@
+package whitelist
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon/core/rawdb"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/log/v3"
+)
+
+type finality[T rawdb.BlockFinality[T]] struct {
+	sync.RWMutex
+	db       kv.RwTx
+	Hash     common.Hash // Whitelisted Hash, populated by reaching out to heimdall
+	Number   uint64      // Number , populated by reaching out to heimdall
+	interval uint64      // Interval, until which we can allow importing
+	doExist  bool
+}
+
+type finalityService interface {
+	IsValidPeer(fetchHeadersByNumber func(number uint64, amount int, skip int, reverse bool) ([]*types.Header, []common.Hash, error)) (bool, error)
+	IsValidChain(currentHeader *types.Header, chain []*types.Header) (bool, error)
+	Get() (bool, uint64, common.Hash)
+	Process(block uint64, hash common.Hash)
+	Purge()
+}
+
+// IsValidPeer checks if the chain we're about to receive from a peer is valid or not
+// in terms of reorgs. We won't reorg beyond the last bor finality submitted to mainchain.
+func (f *finality[T]) IsValidPeer(fetchHeadersByNumber func(number uint64, amount int, skip int, reverse bool) ([]*types.Header, []common.Hash, error)) (bool, error) {
+	// We want to validate the chain by comparing the last finalized block
+	f.RLock()
+
+	doExist := f.doExist
+	number := f.Number
+	hash := f.Hash
+
+	f.RUnlock()
+
+	return isValidPeer(fetchHeadersByNumber, doExist, number, hash)
+}
+
+// IsValidChain checks the validity of chain by comparing it
+// against the local checkpoint entry
+// todo: need changes
+func (f *finality[T]) IsValidChain(currentHeader *types.Header, chain []*types.Header) (bool, error) {
+	// Return if we've received empty chain
+	if len(chain) == 0 {
+		return false, nil
+	}
+
+	res, err := isValidChain(currentHeader, chain, f.doExist, f.Number, f.Hash, f.interval)
+
+	return res, err
+}
+
+func (f *finality[T]) Process(block uint64, hash common.Hash) {
+	f.doExist = true
```

Comment:
Won't we need a lock here?

Reply:
Okay I think we only have a RLock above so won't need in Process while writing but we'll still confirm.